libc-top-half/musl/src/thread/pthread_create.c

```diff
@@ -212,10 +201,15 @@ static void __pthread_exit(void *result)
 #ifdef __wasilibc_unmodified_upstream
 	for (;;) __syscall(SYS_exit, 0);
 #else
-	// __syscall(SYS_exit) would unlock the thread, list
-	// do it manually here
-	__tl_unlock();
 	// Can't use `exit()` here, because it is too high level
+
+	/* On Linux, the thread is created with CLONE_CHILD_CLEARTID,
+	 * and the lock (__thread_list_lock) will be unlocked by kernel when
+	 * this thread terminates.
+	 * See also set_tid_address(2)
+	 *
+	 * In WebAssembly, we leave it to wasi_thread_start instead.
+	 */
```

Comment:
Could we instead call `__tl_unlock` right before `a_store(&self->detach_state, DT_EXITED);`  like we do right be `free(self->map_base);`?  

In the first case we are unlocking right before notifying the joiner that they can call free and in the later case we are unlocking right before we call `free` ourselves, so it seem symmetrical.

(I'm suggesting this so that we can avoid re-implementing `__tl_unlock` in asm if we can).

Reply:
Actually, what was wrong with the single `__tl_unlock` on line 172 that we had previously?

Reply:
> Could we instead call `__tl_unlock` right before `a_store(&self->detach_state, DT_EXITED);` like we do right be `free(self->map_base);`?
> 
> In the first case we are unlocking right before notifying the joiner that they can call free and in the later case we are unlocking right before we call `free` ourselves, so it seem symmetrical.

* the detached case is broken as alex said.
* detached threads do never have joiners.

> 
> (I'm suggesting this so that we can avoid re-implementing `__tl_unlock` in asm if we can).

this PR doesn't re-implement `__tl_unlock`.
it emulates `CLONE_CHILD_CLEARTID`, which the `__tl_lock/unlock/sync` protocol relies on.

Reply:
> Actually, what was wrong with the single `__tl_unlock` on line 172 that we had previously?

* double unlock.
* it unlocks too early and allows joiner to free our stack before we finish on it.

Reply:
I agree the double unlock thing looks like a bug.

But the joiner is not waiting on the tl_lock is it?  The joiner seems to be waiting on `t->detach_state`.  In fact, I don't see the tl_lock referenced at all in `pthread_join.c`.  Maybe I'm missing something?

Reply:
the joiner uses `__td_sync` to sync with the exit.